legacy/builder/print_used_and_not_used_libraries.go

```diff
@@ -29,32 +28,31 @@ type PrintUsedAndNotUsedLibraries struct {
 }
 
 func (s *PrintUsedAndNotUsedLibraries) Run(ctx *types.Context) error {
-	var logLevel string
-	// Print this message as warning when the sketch didn't compile,
-	// as info when we're verbose and not all otherwise
-	if s.SketchError {
-		logLevel = constants.LOG_LEVEL_WARN
-	} else if ctx.Verbose {
-		logLevel = constants.LOG_LEVEL_INFO
-	} else {
+	// Print this message:
+	// - as warning, when the sketch didn't compile
+	// - as info, when verbose is on
+	// - otherwise, output nothing
+	if !s.SketchError && !ctx.Verbose {
 		return nil
 	}
 
-	logger := ctx.GetLogger()
-	libraryResolutionResults := ctx.LibrariesResolutionResults
-
-	for header, libResResult := range libraryResolutionResults {
+	res := ""
+	for header, libResResult := range ctx.LibrariesResolutionResults {
 		if len(libResResult.NotUsedLibraries) == 0 {
 			continue
 		}
-		logger.Fprintln(os.Stdout, logLevel, tr("Multiple libraries were found for \"{0}\""), header)
-		logger.Fprintln(os.Stdout, logLevel, " "+tr("Used: {0}"), libResResult.Library.InstallDir)
+		res += fmt.Sprintln(tr(`Multiple libraries were found for "%[1]s"`, header))
+		res += fmt.Sprintln("  " + tr("Used: %[1]s", libResResult.Library.InstallDir))
 		for _, notUsedLibrary := range libResResult.NotUsedLibraries {
-			logger.Fprintln(os.Stdout, logLevel, " "+tr("Not used: {0}"), notUsedLibrary.InstallDir)
+			res += fmt.Sprintln("  " + tr("Not used: %[1]s", notUsedLibrary.InstallDir))
```

Comment:
```suggestion
			res += fmt.Sprint("\n  " + tr("Not used: %[1]s", notUsedLibrary.InstallDir))
```
The established form of this message is:
```
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-objcopy" -O ihex -R .eeprom "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.elf" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.hex"
Multiple libraries were found for "Foo.h"
  Used: C:\Users\per\Documents\Arduino\libraries\Foo
  Not used: C:\Users\per\Documents\Arduino\libraries\Bar
Using library Foo at version 1.0.7 in folder: C:\Users\per\Documents\Arduino\libraries\Foo
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-size" -A "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.elf"
```

A blank line was introduced after the list by the changes in this PR, causing it to take this form

```
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-objcopy" -O ihex -R .eeprom "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.elf" "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.hex"
Multiple libraries were found for "Foo.h"
  Used: C:\Users\per\Documents\Arduino\libraries\Foo
  Not used: C:\Users\per\Documents\Arduino\libraries\Bar

Using library Foo at version 1.0.7 in folder: C:\Users\per\Documents\Arduino\libraries\Foo
"C:\\Users\\per\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\7.3.0-atmel3.6.1-arduino7/bin/avr-size" -A "C:\\Users\\per\\AppData\\Local\\Temp\\arduino-sketch-A91912F4E8442FE7AB972C142696BA39/FooSketch.ino.elf"
```
I feel that blank line makes an inherently confusing message more difficult for a newcomer to interpret because the "Using library..." line is no longer visually associated with the previous lines.

My suggestion removes that added blank line, restoring it to the previous format.

Reply:
This change:
```diff
-			res += fmt.Sprintln("  " + tr("Not used: %[1]s", notUsedLibrary.InstallDir))
+			res += fmt.Sprint("\n  " + tr("Not used: %[1]s", notUsedLibrary.InstallDir))
```
will output `\nNot used: Foo` instead of `Not used: Foo\n`, so it leaves the cursor at the end of the string, I don't think it's your intended behavior (I think it may behave weirdly if there are multiple libraries).

Anyway, I changed each `Fprintln` with `Sprintln`, 1:1, without adding any extra `\n`... I would expect the same output as before. Maybe the extra blank is coming from somewhere else...

Reply:
>Maybe the extra blank is coming from somewhere else...

it is coming from here:
https://github.com/arduino/arduino-cli/blob/dd7a195fcb85c4fffa189f2754ee999d62361f6f/legacy/builder/print_used_and_not_used_libraries.go#L51-L55

https://github.com/arduino/arduino-cli/blob/dd7a195fcb85c4fffa189f2754ee999d62361f6f/legacy/builder/types/context.go#L255-L273

Reply:
Oh I see now, it's ctx.Info(..) that adds an extra `\n` at the end of the string, will fix this in a moment.